ql/src/java/org/apache/hadoop/hive/ql/exec/repl/util/FileList.java

```diff
@@ -18,158 +18,176 @@
+  public FileList(Path backingFile, HiveConf conf) {
     this.backingFile = backingFile;
     this.conf = conf;
-    if (cacheSize > 0) {
-      // Cache size must be > 0 for this list to be used for the write operation.
-      this.cache = new LinkedBlockingQueue<>(cacheSize);
-      fileListStreamer = new FileListStreamer(cache, backingFile, conf);
-      thresholdPoint = getThreshold(cacheSize);
-      LOG.debug("File list backed by {} can be used for write operation.", backingFile);
-    } else {
-      thresholdHit = true;
+    this.retryMode = false;
+  }
+
+  public void add(String entry) throws IOException {
+    Retryable retryable = buildRetryable();
+    try {
+      retryable.executeCallable((Callable<Void>) ()-> {
+        synchronized (backingFile) {
+          try{
+            if (backingFileWriter == null) {
+              backingFileWriter = initWriter();
+            }
+            backingFileWriter.writeBytes(getEntryWithNewline(entry));
+            backingFileWriter.hflush();
+            LOG.info("Writing entry {} to file list backed by {}", entry, backingFile);
+          } catch (IOException e) {
+            LOG.error("Writing entry {} to file list {} failed, attempting retry.", entry, backingFile, e);
+            this.retryMode = true;
+            close();
+            throw e;
+          }
+        }
+        return null;
+      });
+    } catch (Exception e) {
+      throw new IOException(ErrorMsg.REPL_RETRY_EXHAUSTED.format(e.getMessage()));
     }
   }
 
-  @VisibleForTesting
-  FileList(Path backingFile, FileListStreamer fileListStreamer, LinkedBlockingQueue<String> cache, HiveConf conf) {
-    this.backingFile = backingFile;
-    this.fileListStreamer = fileListStreamer;
-    this.cache = cache;
-    this.conf = conf;
-    thresholdPoint = getThreshold(cache.remainingCapacity());
+  Retryable buildRetryable() {
+    return Retryable.builder()
+            .withHiveConf(conf)
+            .withRetryOnException(IOException.class).build();
   }
 
-  /**
-   * Only add operation is safe for concurrent operations.
-   */
-  public void add(String entry) throws SemanticException {
-    if (thresholdHit && !fileListStreamer.isAlive()) {
-      throw new SemanticException("List is not getting saved anymore to file " + backingFile.toString());
+  // Return the entry ensuring it ends with newline.
+  private String getEntryWithNewline(String entry) {
+    return new StringWriter()
+            .append(entry)
+            .append(System.lineSeparator())
+            .toString();
+  }
+
+  FSDataOutputStream initWriter() throws IOException {
+    if(shouldAppend()) {
+      return getWriterAppendMode(); // append in retry-mode if file has been created already
+    }
+    else {
+      return getWriterCreateMode();
     }
+  }
+
+  boolean shouldAppend() throws IOException {
+    return backingFile.getFileSystem(conf).exists(backingFile) && this.retryMode;
```

Comment:
When do you expect backingFile.getFileSystem(conf).exists(backingFile) as true but this.retryMode is false and what does it mean?

Reply:
In case of bootstrap checkpointing there could be an earlier file lying around from the previous attempt. We rewrite this file in that case.
To distinguish that case from an ongoing retry in the current dump, this flag is introduced.

Reply:
makes sense.